File: controller/article.go
package controller

import (
	"errors"
	"strconv"

	"github.com/labstack/echo"
	"github.com/pwcong/ablog/model"
	"github.com/pwcong/ablog/service"
)

type ArticleController struct {
	Base *BaseController
}

type ArticleForm struct {
	PageSize   int    `query:"page_size"`
	PageNo     int    `query:"page_no"`
	Title      string `json:"title" form:"title"`
	Content    string `json:"content" form:"content"`
	Banner     string `json:"banner" form:"banner"`
	CategoryID uint   `json:"category_id" form:"category_id"`
	TagIDs     []uint `json:"tag_ids" form:"tag_ids"`
}

func (ctx *ArticleController) AddArticle(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	form := new(ArticleForm)
	if err := c.Bind(form); err != nil {
		return err
	}

	article, err := service.AddArticle(form.Title, form.Content, form.Banner, form.CategoryID, form.TagIDs)

	if err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "add article successfully", article)
}

func (ctx *ArticleController) UpdateArticle(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	_id := c.Param("id")

	id, err := strconv.Atoi(_id)
	if err != nil {
		return err
	}

	form := new(ArticleForm)
	if err := c.Bind(form); err != nil {
		return err
	}

	article, err := service.UpdateArticle(uint(id), form.Title, form.Content, form.Banner, form.CategoryID, form.TagIDs)

	if err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "update article successfully", article)
}

func (ctx *ArticleController) GetArticle(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	_id := c.Param("id")

	id, err := strconv.Atoi(_id)
	if err != nil {
		return err
	}

	article, err := service.GetArticle(uint(id))

	if err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "get article successfully", article)
}

func (ctx *ArticleController) DelArticle(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	_id := c.Param("id")
	id, err := strconv.Atoi(_id)
	if err != nil {
		return err
	}

	if err := service.DelArticle(uint(id)); err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "delete article successfully", nil)

}

func (ctx *ArticleController) SearchArticles(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	filter := c.Param("filter")
	if filter != "title" && filter != "content" {
		return errors.New("invalid filter")
	}

	value := c.Param("value")
	form := new(ArticleForm)
	if err := c.Bind(form); err != nil {
		return err
	}

	pageNo, pageSize := ResolvePageParameter(form.PageNo, form.PageSize)
	page, err := service.SearchArticles(filter, value, pageNo, pageSize)

	if err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "search articles successfully", page)
}

func (ctx *ArticleController) GetArticlesByFilter(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	filter := c.Param("filter")
	if filter != "category" && filter != "tag" {
		return errors.New("invalid filter")
	}

	_id := c.Param("id")
	id, err := strconv.Atoi(_id)
	if err != nil {
		return err
	}

	form := new(ArticleForm)
	if err := c.Bind(form); err != nil {
		return err
	}
	pageNo, pageSize := ResolvePageParameter(form.PageNo, form.PageSize)

	var page model.Page
	if filter == "category" {
		page, err = service.GetArticlesByCategoryID(uint(id), pageNo, pageSize)
	} else if filter == "tag" {
		page, err = service.GetArticlesByTagID(uint(id), pageNo, pageSize)
	}

	if err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "get articles successfully", page)
}

func (ctx *ArticleController) GetArticles(c echo.Context) error {

	service := service.ArticleService{Base: ctx.Base.Service}

	form := new(ArticleForm)
	if err := c.Bind(form); err != nil {
		return err
	}
	pageNo, pageSize := ResolvePageParameter(form.PageNo, form.PageSize)

	page, err := service.GetArticles(pageNo, pageSize)
	if err != nil {
		return err
	}

	return BaseResponse(c, true, STATUS_OK, "get articles successfully", page)

}

File: view/ablog-admin/src/network/api/article.js
import request from '../request';

import api from './config';

export function getArticles(page_no, page_size) {
  return request(
    api.article.getArticles.url(page_no, page_size),
    api.article.getArticles.method,
    {},
    {}
  );
}

export function getArticlesByFilter(filter, id, page_no, page_size) {
  return request(
    api.article.getArticlesByFilter.url(filter, id, page_no, page_size),
    api.article.getArticlesByFilter.method,
    {},
    {}
  );
}

export function getArticle(id) {
  return request(
    api.article.getArticle.url(id),
    api.article.getArticle.method,
    {},
    {}
  );
}

export function evaluateArticle(id, score, content) {
  return request(
    api.article.evaluateArticle.url(id),
    api.article.evaluateArticle.method,
    api.article.evaluateArticle.data(score, content),
    {}
  );
}

export function modifyArticle(
  id,
  title,
  content,
  banner,
  category_id,
  tag_ids
) {
  return request(
    api.article.modifyArticle.url(id),
    api.article.modifyArticle.method,
    api.article.modifyArticle.data(
      title,
      content,
      banner,
      category_id,
      tag_ids
    ),
    {}
  );
}

export function addArticle(title, content, banner, category_id, tag_ids) {
  return request(
    api.article.addArticle.url(),
    api.article.addArticle.method,
    api.article.addArticle.data(title, content, banner, category_id, tag_ids),
    {}
  );
}

export function delArticle(id) {
  return request(
    api.article.delArticle.url(id),
    api.article.delArticle.method,
    {},
    {}
  );
}
